### ISSUE TITLE
move_directory refuses to move directory containing broken symlink

### ISSUE THREAD

**Issue by tkkcc** [simongoricar/fs-more#2]
if source directory contains broken symbolic link, move_directory with default option will return Error, is it designed? BTW, fs::rename or coreutils mv won't.

**Comment by simongoricar** [comment]
Yikes, definitely not intended - I'd expect at most an error (but I suppose it's debatable whether that should be the behaviour either).

It might be a week or so before I'll be able to more properly take a look at this, but in the meantime:
- Does move_directory_with_progress do the same?
- Are you sure the call panics (as opposed to an error)? Could you provide the stack trace?
- Are there any other specific circumstances (alongside the broken symlink) that could be a requirement for this?

If the requirements are anything beyond a "create broken symlink + call move_directory", a set of steps to reproduce would also be wonderful.

**Comment by tkkcc** [comment]
Sorry, it's not panic, just PreparationError. and `move_directory_with_progress` give same error.
```
PreparationError(CopyPlanningError(UnableToAccess { path: "/home/bilabila/bin/test/a/non-exist", error: Os { code: 40, kind: FilesystemLoop, message: "Too many levels of symbolic links" } }))
```



Yes, it's debatable for how to handle symbol links, so i thinks it's best to provide option for user. 

In some cases, user want to preserve links as it is even if source link is broken, like https://github.com/nrwl/nx/pull/22548/commits/24565bd28860a944678086de59087385b53fa8f1#diff-663ac081b66b48c93df5278df020d105321ea28dc22e143efcbad9888727d153R28. (i think this can be the default, and it matchs `fs::rename`, and fast).

In other cases, user want to follow links to copy origin file or folder. There can be another option for whether recursively do follow links in symbol link to folder. We need to validate first, once there is a broken link, return error before disk operations, like current `fs-more`.

Just personal thought, :)

---
BTW: I test some cases. I don't think `fs-more` needs to match `mv`.

consider move folder with a broken symbol link to `non exist destination`
1. If move folder without across mount point, `fs-more` will return Error, while `fs::rename` and `mv` won't. It's related to #1 
2. If move folder from ext4 to ntfs, `fs-more` will return Error, while `mv` will preserve broken symlink and give no error
3. If move folder from ext4 to ext4 across mount point, `fs-more` will return Error, `mv` will do incomplete move and print error

**Comment by simongoricar** [comment]
I agree that this should not error by default, and instead preserve the broken symlink. And that it should be configurable, with probably just two options: `keep broken` or `abort with error` (which should happen while preparing the move).

Having more configuration options for what to do with valid symlinks in general would be nice as well - I'm still thinking whether we should go all out on symlink options at this point (which would be quite some work), or just start by handling the broken symlinks (the latter seems like a good way to start).

**Comment by simongoricar** [comment]
While trying to implement this for both copying and moving a directory, an interesting problem has come up for Windows targets. 

On Unix there is a single symlinking function regardless of destination type ([`std::os::unix::fs::symlink`](https://doc.rust-lang.org/stable/std/os/unix/fs/fn.symlink.html)), but the Windows part of the standard library has this function split into [`std::os::windows::fs::symlink_dir`](https://doc.rust-lang.org/stable/std/os/windows/fs/fn.symlink_dir.html) and [`std::os::windows::fs::symlink_file`](https://doc.rust-lang.org/stable/std/os/windows/fs/fn.symlink_file.html).

The issue seems to me that, because the symlink is broken at the time of copy or move, we can't know what type the symlink destination used to be. ~~While I admit I haven't experimented, it seems reasonable to conclude that we can't reliably (re)create broken symlinks? Am I missing something?~~

~~Not sure where to go on from here yet.~~ Actually, the symlink type can be accessed using `is_symlink_dir` / `is_symlink_file` (see [`FileTypeExt`](https://doc.rust-lang.org/stable/std/os/windows/fs/trait.FileTypeExt.html)), even when the link is broken.

**Comment by simongoricar** [comment]
Some updates on the implementation: I've been experimenting a little on the `dev` branch, and symlink options for `copy_directory` and `copy_directory_with_progress` are mostly finished. They consist of the general symlink behaviour option (either preserve or follow), as well as the broken symlink behaviour option (either preserve or abort).

I'm currently thinking on the best way to handle this in `move_directory` and `move_directory_with_progress`. I think it's sensible to only offer broken symlink behaviour options here, but even that will need a bit of thinking. 

The dilemma is that `move_directory` (and `move_directory_with_progress`) have two strategies: a rename or a copy-and-delete. Respecting the broken symlinks option while doing copy-and-delete is trivial, but not so much when renaming: if a rename is possible, but the user indicates they want to abort on broken symbolic links, we would need to scan the entire directory tree just to potentially abort on an invalid symbolic link, which seems massively wasteful.

I haven't decided yet, but I'm currently leaning towards still having a broken symbolic link behaviour option for directory move functions, but very clearly documenting that the broken link is still preserved if the rename strategy is used.

**Comment by simongoricar** [comment]
Alright, so this is resolved as of `v0.7.0`, which will be available on crates.io shortly. Thanks for the report and ideas!

Since there are a few other breaking changes it might be a good idea to check out [`CHANGELOG.md`](https://github.com/simongoricar/fs-more/blob/61a42e97129bd500d84f596113ce9d5b20400197/CHANGELOG.md), but in short, there are two new options: `symlink_behaviour` and `broken_symlink_behaviour`, available for both directory copy and move functions. They control whether to follow (dereference) or preserve symbolic links, and more specifically, whether to preserve broken symbolic links or abort when encountering them. 

The defaults are to preserve symbolic links, including broken ones.

(As far as my previous comment goes, I ended up restructuring the way copy-and-delete options are available for directory moves. The documentation now makes it very clear that certain options only apply to that move strategy.)
